src/render/optixdenoiser.cpp

```diff
@@ -0,0 +1,391 @@
+                                          bool temporal) {
+    OptixDeviceContext context = jit_optix_context();
+    OptixDenoiserModelKind model_kind = temporal
+                                            ? OPTIX_DENOISER_MODEL_KIND_TEMPORAL
+                                            : OPTIX_DENOISER_MODEL_KIND_HDR;
+    jit_optix_check(
+        optixDenoiserCreate(context, model_kind, &m_options, &m_denoiser));
+
+    OptixDenoiserSizes sizes = {};
+    jit_optix_check(optixDenoiserComputeMemoryResources(
+        m_denoiser, input_size.x(), input_size.y(), &sizes));
+
+    CUstream stream = jit_cuda_stream();
+    m_state_size = sizes.stateSizeInBytes;
+    m_state = jit_malloc(AllocType::Device, m_state_size);
+    m_scratch_size = sizes.withoutOverlapScratchSizeInBytes;
+    m_scratch = jit_malloc(AllocType::Device, m_scratch_size);
+    jit_optix_check(optixDenoiserSetup(m_denoiser, stream, input_size.x(),
+                                       input_size.y(), m_state, m_state_size,
+                                       m_scratch, m_scratch_size));
+    m_hdr_intensity = jit_malloc(AllocType::Device, sizeof(float));
+}
+
+MI_VARIANT
+void OptixDenoiser<Float, Spectrum>::validate_input(
+    const TensorXf &noisy, const TensorXf *albedo, const TensorXf *normals,
+    const TensorXf *flow, const TensorXf *previous_denoised) const {
+    if ((albedo == nullptr) && m_options.guideAlbedo)
+        Throw("The denoiser was created with albedo guiding enabled. An albedo "
+              "layer must be specified!");
+    if ((normals == nullptr) && m_options.guideNormal)
+        Throw("The denoiser was created with normals guiding enabled. A normal"
+              "layer must be specified!");
+    if (((flow == nullptr) || (previous_denoised == nullptr)) && m_temporal)
+        Throw("The denoiser was created with temporal denoising enabled. Both "
+              "the optical flow and the denoised previous frame must be "
+              "specified!");
+
+    auto check_resolution = [](const TensorXf &tensor,
+                               const ScalarVector2u &expected_size) {
+        if (expected_size.x() != tensor.shape(1) ||
+            expected_size.y() != tensor.shape(0))
+            Throw(
+                "The denoiser was created for inputs of size %u x %u (width x "
+                "height). At least one of the input arguments does not have "
+                "this size. You must create a new denoiser object for inputs "
+                "of different sizes!",
+                expected_size.x(), expected_size.y());
+    };
+    check_resolution(noisy, m_input_size);
+    if (m_options.guideAlbedo)
+        check_resolution(*albedo, m_input_size);
+    if (m_options.guideNormal)
+        check_resolution(*normals, m_input_size);
+    if (m_temporal) {
+        check_resolution(*flow, m_input_size);
+        check_resolution(*previous_denoised, m_input_size);
+    }
```

Comment:
Skip the if statements below by making the lambda do nothing when the input pointer is nullptr.

Reply:
:+1: